invenio_rdm_records/services/components.py

```diff
@@ -144,22 +146,73 @@ def new_version(self, identity, draft=None, record=None, **kwargs):
 class ExternalPIDsComponent(ServiceComponent):
     """Service component for pids."""
 
+    def get_provider(self, scheme):
+        """Process a given PID with its provider."""
+        provider = self.service.config.pids_providers.get(scheme)
+
+        if not provider:
+            raise ValidationError(
+                _(f"No provider configured for PID type {scheme}"),
+                field_name="pids"
+            )
+
+        return provider
+
+    def _validate_pids(self, pids):
+        """Validate an iterator of pids."""
+        for scheme, pid_attrs in pids.items():
+            self.get_provider(scheme).validate(pid_attrs=pid_attrs)
+
     def create(self, identity, data=None, record=None, **kwargs):
         """Inject parsed pids to the record."""
-        record.pids = data.get('pids', {})
+        pids = data.get('pids', {})
+        self._validate_pids(pids)
+        record.pids = pids
 
     def update_draft(self, identity, data=None, record=None, **kwargs):
         """Inject parsed pids to the record."""
-        record.pids = data.get('pids', {})
+        pids = data.get('pids', {})
+        self._validate_pids(pids)
+        record.pids = pids
 
     def publish(self, identity, draft=None, record=None, **kwargs):
         """Update draft pids."""
-        record.pids = draft.get('pids', {})
+        pids = draft.get('pids', {})
+
+        for scheme, pid_attrs in pids.items():
+            provider = self.get_provider(scheme)
+            provider.validate(pid_attrs=pid_attrs)
+
+            identifier_value = pid_attrs.get("identifier")
+            if not identifier_value:
+                pid = provider.create()
+                pid_attrs["identifier"] = pid.value
+                pids[scheme] = pid_attrs
+
+            elif not provider.is_external():
```

Comment:
Here it might be that we have the `identifier` because it was reserved and retrieved in the form via btn Get DOI  now, or maybe I am wrong...

Reply:
Im not sure I get this one, then in line 198 we do a get if this is the case so we have the `PersistentIdentifier` instance so we can actually check stuff on it.